src/protocol/sort/reshare.rs

```diff
@@ -0,0 +1,138 @@
+
+/// A message sent by each helper when they've reshared their own shares
+#[derive(Debug, Serialize, Deserialize, Default)]
+pub struct PartValue<F> {
+    part: F,
+}
+/// Reshare(i, [x])
+// This implements reshare algorithm of "Efficient Secure Three-Party Sorting Protocol with an Honest Majority" at communication cost of 2R.
+// Input: Pi-1 and Pi+1 knows their secret shares
+// Output: At the end of the protocol, [all 3 helpers receive renewed shares of x with randomness added
+#[derive(Debug)]
+pub struct Reshare<F> {
+    input: Replicated<F>,
+}
+
+impl<F: Field> Reshare<F> {
+    #[allow(dead_code)]
+    pub fn new(input: Replicated<F>) -> Self {
+        Self { input }
+    }
+
+    /// Steps
+    #[embed_doc_image("reshare", "images/sort/reshare.png")]
+    /// 1. While calculating for helper, we call PRSS to get randoms for helper who needs to receive reshares (say `rand_left`, `rand_right`)
+    ///    `to_helper.left` knows `rand_left` (named r1) and `to_helper.right` knows `rand_right` (named r0)
+    /// 2. `to_helper.left` calculates part1 = (input.0 + input.1) - r1
+    ///    `to_helper.right` calculates part2 = (input.0 - r0)
+    /// 3. `to_helper.left` and `to_helper.right` exchange their calculated shares
+    /// 4. Everyone sets their shares  
+    ///    `to_helper.left`  = (part1 + part2, `rand_left`)  = (part1 + part2, r1)
+    ///    `to_helper`       = (`rand_left`, `rand_right`)     = (r0, r1)
+    ///    `to_helper.right` = (`rand_right`, part1 + part2) = (r0, part1 + part2)
+    #[allow(dead_code)]
+    pub async fn execute<M: Mesh, G: Gateway<M, S>, S: Step + SpaceIndex>(
+        self,
+        ctx: &ProtocolContext<'_, G, S>,
+        record_id: RecordId,
+        step: S,
+        to_helper: Identity,
+    ) -> Result<Replicated<F>, BoxError> {
+        let mut channel = ctx.gateway.get_channel(step);
+        let prss = &ctx.participant[step];
+        let (r0, r1) = prss.generate_fields(record_id.into());
+
+        let inputs = self.input.as_tuple();
+        // `to_helper.left` calculates part1 = (input.0 + input.1) - r1 and sends part1 to `to_helper.right`
+        if channel.identity() == to_helper.peer(Direction::Left) {
+            let part1 = inputs.0 + inputs.1 - r1;
+            channel
+                .send(
+                    to_helper.peer(Direction::Right),
+                    record_id,
+                    PartValue { part: part1 },
+                )
+                .await?;
+
+            // Sleep until `to_helper.left` sends us their part2 value
+            let PartValue { part: part2 } = channel
+                .receive(channel.identity().peer(Direction::Left), record_id)
+                .await?;
+
+            Ok(Replicated::new(part1 + part2, r1))
+        } else if channel.identity() == to_helper.peer(Direction::Right) {
+            //  `to_helper.right` calculates part2 = (input.0 - r0) and sends it to `to_helper.left`
+            let part2 = inputs.0 - r0;
```

Comment:
Pretty sure you mean "input.2" on both the commend and the next line.

Reply:
From helper POV, there is nothing called inputs.2 since it has two shares inputs.0 and inputs.1
This comment is to align with what is coded. Let me try to make it clearer